.gitlab/llnl-ci.yml

```diff
@@ -1,79 +1,132 @@
+    - spack config add concretizer:reuse:false
+    - spack config add config:db_lock_timeout:300
+    - spack compiler find
+    - spack add $COMPILER %gcc@4.9.3 $TARGET_ARCH_OPT
+    - spack install || (spack env deactivate && spack env remove $SPACK_ENV_NAME -y && exit 1)
+    - spack compiler add $(spack location -i $COMPILER)
+    - *finalize
 
-Spack Build rhel7-ppc64le:
-  stage: build rhel7-ppc64le
-  variables:
-    HOST : "lassen"
-    SPACK_ENV_NAME : "rhel7-ppc64le"
-    COMPILER : "gcc@9.4.0"
+spack-update:
+  stage: update
   tags:
-    - $HOST
+    - quartz
     - shell
-  script: *build_script
+  script:
+    - cd $SPACK_WORK_DIR/spack/
+    - git checkout HEAD^ .
+    - git reset --hard HEAD
+    - git pull --ff
+    - *finalize
 
-Spack Build rhel7-cascadelake:
-  stage: build rhel7-cascadelake
-  variables:
-    HOST : "ruby"
-    SPACK_ENV_NAME : "rhel7-cascadelake"
-    COMPILER : "gcc@8.2.0"
+spack-build:
+  stage: build
+  needs:
+    - job: spack-update
+  parallel:
+    matrix:
+      - HOST: [ "quartz" ]
+        SPACK_ENV_NAME: [ "rhel7-broadwell" ]
+        COMPILER: [ "gcc@8.1.0" ]
+      - HOST: [ "lassen" ]
+        SPACK_ENV_NAME: [ "rhel7-ppc64le" ]
+        COMPILER: [ "gcc@9.4.0" ]
+      - HOST: [ "ruby" ]
+        SPACK_ENV_NAME: [ "rhel7-cascadelake" ]
+        COMPILER: [ "gcc@8.2.0" ]
   tags:
     - $HOST
     - shell
-  script: *build_script
+  script:
+    - . $SPACK_WORK_DIR/spack/share/spack/setup-env.sh
+    - for i in {1..3}; do
+        spack env activate $SPACK_ENV_NAME && break || sleep 5;
+      done
+    - spack uninstall -y exaworks%$COMPILER py-pytest%$COMPILER || true
+    - spack add exaworks%$COMPILER ^python@3.9 py-pytest%$COMPILER
+    - spack concretize -f
+    - spack install
+    - *finalize
 
-Spack Test:
-  variables:
-    LLNL_SLURM_SCHEDULER_PARAMETERS: "--nodes=1 -t 30"
+sdk-test:
   stage: test
+  needs:
+    - job: spack-build
   before_script:
     - whoami
   parallel:
     matrix:
-      - HOST : ["quartz"]
-        SPACK_ENV_NAME : ["rhel7-broadwell"]
-      - HOST : ["lassen"]
-        SPACK_ENV_NAME : ["rhel7-ppc64le"]
-      - HOST : ["ruby"]
-        SPACK_ENV_NAME : ["rhel7-cascadelake"]
+      - HOST: [ "quartz" ]
+        SPACK_ENV_NAME: [ "rhel7-broadwell" ]
+        LLNL_SLURM_SCHEDULER_PARAMETERS: [ "--nodes=1 -t 30" ]
+      - HOST: [ "lassen" ]
+        SPACK_ENV_NAME: [ "rhel7-ppc64le" ]
+        LLNL_LSF_SCHEDULER_PARAMETERS: [ "-nnodes 1 -W 30" ]
+        LSB_JOB_STARTER: [ "ENVIRONMENT=BATCH /usr/tcetmp/bin/bsub_job_starter %USRCMD" ]
+        ENVIRONMENT: [ "BATCH" ]
+      - HOST: [ "ruby" ]
+        SPACK_ENV_NAME: [ "rhel7-cascadelake" ]
+        LLNL_SLURM_SCHEDULER_PARAMETERS: [ "--nodes=1 -t 30" ]
   tags:
-    - batch
     - $HOST
+    - batch
   script:
     - . $SPACK_WORK_DIR/spack/share/spack/setup-env.sh
-    - spack env activate $SPACK_ENV_NAME
+    - for i in {1..3}; do
+        spack env activate $SPACK_ENV_NAME && break || sleep 5;
```

Comment:
What is the point of this loop?

Reply:
we have 3 parallel jobs here and each of them uses the same spack setup, it happened they were interfering with each other, thus here 3 attempts to activate a corresponding spack env